Source/Core/Common/BitFieldView.h

```diff
@@ -0,0 +1,788 @@
+// modification, are permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this
+//    list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+//    this list of conditions and the following disclaimer in the documentation
+//    and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its
+//    contributors may be used to endorse or promote products derived from
+//    this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// BitFieldView.h - A C++17 library
+// TODO: C++20 concepts sound like a good replacement for all of the type sanity static assertions.
+
+#pragma once
+
+#include <cstddef>      // std::size_t, std::int32_t, std::uint32_t, std::int64_t, std::uint64_t
+#include <iterator>     // std::input_iterator_tag, std::output_iterator_tag
+#include <limits>       // std::numeric_limits
+#include <type_traits>  // std::make_unsigned, std::make_signed, std::is_integral, std::is_enum
+
+#include "Common/Assert.h"    // DEBUG_ASSERT
+#include "Common/BitUtils.h"  // Common::BitCast
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Assertation helper classes
+
+template <typename host_t>
+struct AssertHostTypeIsSane
+{
+  AssertHostTypeIsSane() = default;
+  static_assert(std::is_integral_v<host_t> || std::is_enum_v<host_t>,
+                "Given host type is not sane.  You must define a partial specialization for it.");
+};
+template <typename field_t>
+struct AssertFieldTypeIsSane
+{
+  AssertFieldTypeIsSane() = default;
+  static_assert(std::is_integral_v<field_t> || std::is_enum_v<field_t>,
+                "Given field type is not sane.  It must be an integral or enumerated type.");
+};
+template <typename field_t, std::size_t start, std::size_t width, typename host_t>
+struct AssertBitFieldFixedIsSane
+{
+  AssertBitFieldFixedIsSane() = default;
+  static_assert(start <= 8 * sizeof(host_t), "BitField out of range");
```

Comment:
```suggestion
  static_assert(start <= CHAR_BIT * sizeof(host_t), "BitField out of range");
```

You will want `<climits>`. Ditto for all other instances.

Reply:
Perhaps instead consider: `Common::BitSize<host_t>()`.